### PR TITLE
[core] AnimateDiff SparseCtrl

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Adds support for the long awaited AnimateDiff SparseCtrl :rocket:

Paper: https://arxiv.org/abs/2311.16933
Project page: https://guoyww.github.io/projects/SparseCtrl/
Code: https://github.com/guoyww/AnimateDiff/

SparseCtrl RGB: https://huggingface.co/a-r-r-o-w/animatediff-sparsectrl-rgb
SparseCtrl Scribble: https://huggingface.co/a-r-r-o-w/animatediff-sparsectrl-scribble
SparseCtrl Motion Lora: https://huggingface.co/a-r-r-o-w/animatediff-motion-lora-v1-5-3

TODO:
- [x] convert scribble variant, lora
- [x] verify outputs with original implementation
- [x] show example results
- [x] tests
- [x] docs
- [x] move converted checkpoints to https://huggingface.co/guoyww

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

@DN6 @sayakpaul

### CODE REVIEW

**Comment by a-r-r-o-w** [comment]
Posting some example snippets and results. All input images/videos used in the code can be found [here](https://drive.google.com/drive/folders/1zAnP2N869EuSWQTCXWXcj_ozCe8K_Nhd?usp=sharing)

**Scribble - multiple image interpolation**

<details>
<summary> Code </summary>

```py
import torch

from diffusers import AnimateDiffSparseControlNetPipeline
from diffusers.models import AutoencoderKL, MotionAdapter, SparseControlNetModel
from diffusers.schedulers import DPMSolverMultistepScheduler
from diffusers.utils import export_to_gif, load_image


model_id = "SG161222/Realistic_Vision_V5.1_noVAE"
motion_adapter_id = "guoyww/animatediff-motion-adapter-v1-5-3"
controlnet_id = "a-r-r-o-w/animatediff-sparsectrl-scribble"
vae_id = "stabilityai/sd-vae-ft-mse"
lora_adapter_id = "a-r-r-o-w/animatediff-motion-lora-v1-5-3"
device = "cuda:1"

motion_adapter = MotionAdapter.from_pretrained(motion_adapter_id, torch_dtype=torch.float16).to(device)
controlnet = SparseControlNetModel.from_pretrained(controlnet_id, torch_dtype=torch.float16).to(device)
vae = AutoencoderKL.from_pretrained(vae_id, torch_dtype=torch.float16).to(device)
scheduler = DPMSolverMultistepScheduler.from_pretrained(
    model_id,
    subfolder="scheduler",
    beta_schedule="linear",
    algorithm_type="dpmsolver++",
    use_karras_sigmas=True,
)
pipe = AnimateDiffSparseControlNetPipeline.from_pretrained(
    model_id,
    motion_adapter=motion_adapter,
    controlnet=controlnet,
    vae=vae,
    scheduler=scheduler,
    torch_dtype=torch.float16,
).to(device)
pipe.load_lora_weights(lora_adapter_id, adapter_name="motion_lora")
pipe.fuse_lora(lora_scale=1.0)

prompt = "an aerial view of a cyberpunk city, night time, neon lights, masterpiece, high quality"
negative_prompt = "low quality, worst quality, letterboxed"

image_files = ["scribble-1.png", "scribble-2.png", "scribble-3.png"]
condition_frame_indices = [0, 8, 15]
images = [load_image(img_file) for img_file in image_files]

video = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    conditioning_frames=images,
    height=512,
    width=512,
    num_inference_steps=25,
    guidance_scale=10,
    controlnet_conditioning_scale=1.0,
    controlnet_frame_indices=condition_frame_indices,
    generator=torch.Generator().manual_seed(1337),
).frames[0]
export_to_gif(video, "results/animatediff_sparsectrl_scribble_multiple.gif")
```
</details>

<table align=center>
<tr><td></td></tr>
</table>

**Scribble - canny inputs + low frame rate input**

<details>
<summary> Code </summary>

```py
import requests
from io import BytesIO

import imageio
import torch
from PIL import Image

from controlnet_aux.processor import Processor
from diffusers import AnimateDiffSparseControlNetPipeline
from diffusers.models import AutoencoderKL, MotionAdapter, SparseControlNetModel
from diffusers.schedulers import DPMSolverMultistepScheduler
from diffusers.utils import export_to_gif, load_image


def load_video(file_path: str):
    images = []

    if file_path.startswith(('http://', 'https://')):
        # If the file_path is a URL
        response = requests.get(file_path)
        response.raise_for_status()
        content = BytesIO(response.content)
        vid = imageio.get_reader(content)
    else:
        # Assuming it's a local file path
        vid = imageio.get_reader(file_path)

    for frame in vid:
        pil_image = Image.fromarray(frame)
        images.append(pil_image)

    return images


model_id = "SG161222/Realistic_Vision_V5.1_noVAE"
motion_adapter_id = "guoyww/animatediff-motion-adapter-v1-5-3"
controlnet_id = "a-r-r-o-w/animatediff-sparsectrl-scribble"
vae_id = "stabilityai/sd-vae-ft-mse"
lora_adapter_id = "a-r-r-o-w/animatediff-motion-lora-v1-5-3"
device = "cuda:1"

motion_adapter = MotionAdapter.from_pretrained(motion_adapter_id, torch_dtype=torch.float16).to(device)
controlnet = SparseControlNetModel.from_pretrained(controlnet_id, torch_dtype=torch.float16).to(device)
vae = AutoencoderKL.from_pretrained(vae_id, torch_dtype=torch.float16).to(device)
scheduler = DPMSolverMultistepScheduler.from_pretrained(
    model_id,
    subfolder="scheduler",
    beta_schedule="linear",
    algorithm_type="dpmsolver++",
    use_karras_sigmas=True,
)
pipe = AnimateDiffSparseControlNetPipeline.from_pretrained(
    model_id,
    motion_adapter=motion_adapter,
    controlnet=controlnet,
    vae=vae,
    scheduler=scheduler,
    torch_dtype=torch.float16,
).to(device)
pipe.load_lora_weights(lora_adapter_id, adapter_name="motion_lora")
pipe.fuse_lora(lora_scale=1.0)

processor = Processor("canny")

prompt = "1girl, dancing, bustling metropolis, anime, vibrant colors"
negative_prompt = "low quality, worst quality"
frames = load_video("input.gif")
assert len(frames) >= 16

# get low frame rate input
condition_frame_indices = [1, 4, 8, 12, 15]
conditioning_frames = [processor(frames[i]) for i in condition_frame_indices]

video = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    conditioning_frames=conditioning_frames,
    height=768,
    width=512,
    num_inference_steps=25,
    guidance_scale=10,
    controlnet_conditioning_scale=1.0,
    controlnet_frame_indices=condition_frame_indices,
    generator=torch.Generator().manual_seed(1337),
).frames[0]
export_to_gif(video, "results/animatediff_sparsectrl_scribble_low_frame_rate.gif")
```
</details>

<table align=center>
<tr><td></td></tr>
</table>

**Scribble - IP Adapter**

<details>
<summary> Code </summary>

```py
import torch

from diffusers import AnimateDiffSparseControlNetPipeline
from diffusers.models import AutoencoderKL, MotionAdapter, SparseControlNetModel
from diffusers.schedulers import DPMSolverMultistepScheduler
from diffusers.utils import export_to_gif, load_image


model_id = "SG161222/Realistic_Vision_V5.1_noVAE"
motion_adapter_id = "guoyww/animatediff-motion-adapter-v1-5-3"
controlnet_id = "a-r-r-o-w/animatediff-sparsectrl-scribble"
vae_id = "stabilityai/sd-vae-ft-mse"
lora_adapter_id = "a-r-r-o-w/animatediff-motion-lora-v1-5-3"
device = "cuda:1"

motion_adapter = MotionAdapter.from_pretrained(motion_adapter_id, torch_dtype=torch.float16).to(device)
controlnet = SparseControlNetModel.from_pretrained(controlnet_id, torch_dtype=torch.float16).to(device)
vae = AutoencoderKL.from_pretrained(vae_id, torch_dtype=torch.float16).to(device)
scheduler = DPMSolverMultistepScheduler.from_pretrained(
    model_id,
    subfolder="scheduler",
    beta_schedule="linear",
    algorithm_type="dpmsolver++",
    use_karras_sigmas=True,
)
pipe = AnimateDiffSparseControlNetPipeline.from_pretrained(
    model_id,
    motion_adapter=motion_adapter,
    controlnet=controlnet,
    vae=vae,
    scheduler=scheduler,
    torch_dtype=torch.float16,
).to(device)
pipe.load_lora_weights(lora_adapter_id, adapter_name="motion_lora")
pipe.fuse_lora(lora_scale=1.0)
pipe.load_ip_adapter("h94/IP-Adapter", subfolder="models", weight_name="ip-adapter_sd15.bin")
pipe.set_ip_adapter_scale({
    "down": {"block_2": [0.0, 1.0]},
    "up": {"block_0": [0.0, 1.0, 0.0]},
})

ip_adapter_image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/0052a70beed5bf71b92610a43a52df6d286cd5f3/diffusers/rabbit.jpg")

image_files = ["scribble-1.png", "scribble-2.png", "scribble-3.png"]
condition_frame_indices = [0, 8, 15]
images = [load_image(img_file) for img_file in image_files]

video = pipe(
    prompt="drone view of buildings",
    negative_prompt="low quality, worst quality",
    conditioning_frames=images,
    ip_adapter_image=ip_adapter_image,
    height=512,
    width=512,
    num_inference_steps=25,
    guidance_scale=10,
    controlnet_conditioning_scale=1.0,
    controlnet_frame_indices=condition_frame_indices,
    generator=torch.Generator().manual_seed(1337),
).frames[0]
export_to_gif(video, "results/animatediff_sparsectrl_scribble_ipadapter.gif")
```
</details>

<table align=center>
<tr><td></td></tr>
</table>

**RGB - low frame rate input**

<details>
<summary> Code </summary>

```py
import torch

from diffusers import AnimateDiffSparseControlNetPipeline
from diffusers.models import AutoencoderKL, MotionAdapter, SparseControlNetModel
from diffusers.schedulers import DPMSolverMultistepScheduler
from diffusers.utils import export_to_gif, load_image


model_id = "SG161222/Realistic_Vision_V5.1_noVAE"
motion_adapter_id = "guoyww/animatediff-motion-adapter-v1-5-3"
controlnet_id = "a-r-r-o-w/animatediff-sparsectrl-rgb"
vae_id = "stabilityai/sd-vae-ft-mse"
lora_adapter_id = "a-r-r-o-w/animatediff-motion-lora-v1-5-3"
device = "cuda:1"

motion_adapter = MotionAdapter.from_pretrained(motion_adapter_id, torch_dtype=torch.float16).to(device)
controlnet = SparseControlNetModel.from_pretrained(controlnet_id, torch_dtype=torch.float16).to(device)
vae = AutoencoderKL.from_pretrained(vae_id, torch_dtype=torch.float16).to(device)
scheduler = DPMSolverMultistepScheduler.from_pretrained(
    model_id,
    subfolder="scheduler",
    beta_schedule="linear",
    algorithm_type="dpmsolver++",
    use_karras_sigmas=True,
)
pipe = AnimateDiffSparseControlNetPipeline.from_pretrained(
    model_id,
    motion_adapter=motion_adapter,
    controlnet=controlnet,
    vae=vae,
    scheduler=scheduler,
    torch_dtype=torch.float16,
).to(device)
pipe.load_lora_weights(lora_adapter_id, adapter_name="motion_lora")

prompt = "two people holding hands in a field with wind turbines in the background"
negative_prompt = "low quality, worst quality, letterboxed"

image_files = ["lfps-1.png", "lfps-2.png", "lfps-3.png", "lfps-4.png"]
condition_frame_indices = [0, 5, 10, 15]
images = [load_image(img_file) for img_file in image_files]

video = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    conditioning_frames=images,
    height=256,
    width=384,
    num_inference_steps=25,
    guidance_scale=7,
    controlnet_conditioning_scale=1.0,
    generator=torch.Generator().manual_seed(42),
).frames[0]
export_to_gif(video, "results/animatediff_sparsectrl_rgb_lfps.gif")
```
</details>

<table align=center>
<tr><td></td></tr>
</table>

- RGB - single image

<details>
<summary> Code </summary>

```py
import torch

from diffusers import AnimateDiffSparseControlNetPipeline
from diffusers.models import AutoencoderKL, MotionAdapter, SparseControlNetModel
from diffusers.schedulers import DPMSolverMultistepScheduler
from diffusers.utils import export_to_gif, load_image


model_id = "SG161222/Realistic_Vision_V5.1_noVAE"
motion_adapter_id = "guoyww/animatediff-motion-adapter-v1-5-3"
controlnet_id = "a-r-r-o-w/animatediff-sparsectrl-rgb"
vae_id = "stabilityai/sd-vae-ft-mse"
lora_adapter_id = "a-r-r-o-w/animatediff-motion-lora-v1-5-3"
device = "cuda:1"

motion_adapter = MotionAdapter.from_pretrained(motion_adapter_id, torch_dtype=torch.float16).to(device)
controlnet = SparseControlNetModel.from_pretrained(controlnet_id, torch_dtype=torch.float16).to(device)
vae = AutoencoderKL.from_pretrained(vae_id, torch_dtype=torch.float16).to(device)
scheduler = DPMSolverMultistepScheduler.from_pretrained(
    model_id,
    subfolder="scheduler",
    beta_schedule="linear",
    algorithm_type="dpmsolver++",
    use_karras_sigmas=True,
)
pipe = AnimateDiffSparseControlNetPipeline.from_pretrained(
    model_id,
    motion_adapter=motion_adapter,
    controlnet=controlnet,
    vae=vae,
    scheduler=scheduler,
    torch_dtype=torch.float16,
).to(device)
pipe.load_lora_weights(lora_adapter_id, adapter_name="motion_lora")

prompt = "closeup face photo of man in black clothes, night city street, bokeh, fireworks in background"
negative_prompt = "low quality, worst quality"
image = load_image("firework.png")

video = pipe(
    prompt=prompt,
    negative_prompt=negative_prompt,
    conditioning_frames=image,
    height=512,
    width=512,
    num_inference_steps=25,
    guidance_scale=8.5,
    controlnet_conditioning_scale=1.0,
    generator=torch.Generator().manual_seed(42),
).frames[0]
export_to_gif(video, "results/animatediff_sparsectrl_rgb_single.gif")
```
</details>

<table align=center>
<tr><td></td></tr>
</table>

I haven't been able to fully replicate SparseCtrl RGB at the moment, I think, as can be seen from the weird outputs compared to the original demos. However, if you look at the original demos closely, we can observe this sudden flicker behaviour too, which could mean the implementation is okay but poor model quality. There are some reports about this so it might be quite possible that this is expected, and what we're observing is the RGB model not being too good: https://github.com/guoyww/AnimateDiff/issues/274 and https://github.com/guoyww/AnimateDiff/issues/278.

File: scripts/convert_animatediff_motion_lora_to_diffusers.py [hunk 1]
```diff
@@ -51,4 +59,12 @@ def get_args():
...
-    save_file(output_dict, f"{args.output_path}/diffusion_pytorch_model.safetensors")
+    os.makedirs(args.output_path, exist_ok=True)
+
+    filepath = os.path.join(args.output_path, "diffusion_pytorch_model.safetensors")
+    save_file(output_dict, filepath)
+
+    if args.push_to_hub:
+        api = HfApi()
+        repo_id = api.create_repo(args.output_path, exist_ok=True).repo_id
+        api.upload_folder(repo_id=repo_id, folder_path=args.output_path, repo_type="model")
```
Comment (sayakpaul) [inline scripts/convert_animatediff_motion_lora_to_diffusers.py:70]: We can directly use `create_repo()` and `upload_folder()` from `huggingface_hub`. We don't need to use `HfApi()` here.

File: docs/source/en/api/models/controlnet_sparsectrl.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+```py
+import torch
+from diffusers import SparseControlNetModel
+
+# fp32 variant in float16
+controlnet = SparseControlNetModel.from_pretrained("a-r-r-o-w/animatediff-sparsectrl-rgb", torch_dtype=torch.float16)
+
+# fp16 variant
+controlnet = SparseControlNetModel.from_pretrained("a-r-r-o-w/animatediff-sparsectrl-rgb", variant="fp16", torch_dtype=torch.float16)
+```
```
Comment (sayakpaul) [inline docs/source/en/api/models/controlnet_sparsectrl.md:35]: Is this a full-fledged example? I would prefer an end-to-end example here. 

Also, I think we should move the "animatediff-sparsectrl-rgb" checkpoints to their appropriate orgs.

File: src/diffusers/models/controlnet_sparsectrl.py [hunk 3]
```diff
@@ -0,0 +1,790 @@
...
+                module.set_attention_slice(slice_size.pop())
+
+            for child in module.children():
+                fn_recursive_set_attention_slice(child, slice_size)
+
+        reversed_slice_size = list(reversed(slice_size))
+        for module in self.children():
+            fn_recursive_set_attention_slice(module, reversed_slice_size)
+
+    def _set_gradient_checkpointing(self, module, value: bool = False) -> None:
```
Comment (sayakpaul) [inline src/diffusers/models/controlnet_sparsectrl.py:595]: But we are not doing any checkpointing in the `forward()`. Perhaps that needs to be implemented?

File: src/diffusers/models/controlnet_sparsectrl.py [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+        resnet_time_scale_shift: str = "default",
+        projection_class_embeddings_input_dim: Optional[int] = None,
+        controlnet_conditioning_channel_order: str = "rgb",
+        conditioning_embedding_out_channels: Optional[Tuple[int, ...]] = (16, 32, 96, 256),
+        global_pool_conditions: bool = False,
+        addition_embed_type_num_heads: int = 64,
+        motion_max_seq_length: int = 32,
+        motion_num_attention_heads: int = 8,
+        concat_conditioning_mask: bool = True,
+        use_simplified_condition_embedding: bool = True,
```
Comment (sayakpaul) [inline src/diffusers/models/controlnet_sparsectrl.py:215]: Do we need to all the config variable here? Can we reduce some of them?

File: src/diffusers/models/controlnet_sparsectrl.py [hunk 5]
```diff
@@ -0,0 +1,790 @@
...
+
+        return SparseControlNetOutput(
+            down_block_res_samples=down_block_res_samples, mid_block_res_sample=mid_block_res_sample
+        )
+
+
+def zero_module(module: nn.Module) -> nn.Module:
+    for p in module.parameters():
+        nn.init.zeros_(p)
+    return module
```
Comment (sayakpaul) [inline src/diffusers/models/controlnet_sparsectrl.py:790]: If this copied, let's add a copy statement.

**Review by sayakpaul — COMMENTED** [review]
Thanks! I left high-level comments. For a deeper review, I will defer to @DN6. I can then review one more time after that.

File: docs/source/en/api/models/controlnet_sparsectrl.md (hunk 2 above)
Comment (a-r-r-o-w) [reply docs/source/en/api/models/controlnet_sparsectrl.md:35]: > Is this a full-fledged example? I would prefer an end-to-end example here.

This is just the model documentation, no? I've noticed we have simple examples like this for model docs and a full-fledged example in pipeline docs. But okay to add a full example here as well.

> Also, I think we should move the "animatediff-sparsectrl-rgb" checkpoints to their appropriate orgs.

Yes, will need help with this. There are 3 checkpoints that will need to be moved

File: src/diffusers/models/controlnet_sparsectrl.py (hunk 3 above)
Comment (a-r-r-o-w) [reply src/diffusers/models/controlnet_sparsectrl.py:595]: Will take a look at this. Some of the code here is remnants from copying the `ControlNetModel`

Comment (a-r-r-o-w) [reply src/diffusers/models/controlnet_sparsectrl.py:595]: This seems correct and we don't have to make any changes in the forward here if I follow the implementation of `ControlNetModel` as the source of truth. The DownBlockMotion, CrossAttnDownBlockMotion and UNetMidBlock2DCrossAttn have it implemented.

File: docs/source/en/api/models/controlnet_sparsectrl.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
+<!-- Copyright 2024 The HuggingFace Team and The InstantX Team. All rights reserved.
```
Comment (DN6) [inline docs/source/en/api/models/controlnet_sparsectrl.md:1]: Change license here to just HF Team.

File: src/diffusers/models/controlnet.py [hunk 7]
```diff
@@ -830,7 +830,6 @@ def forward(
                 sample = self.mid_block(sample, emb)
 
         # 5. Control net blocks
-
```
Comment (DN6) [inline src/diffusers/models/controlnet.py:833]: make style or accident?

File: src/diffusers/models/controlnet_sparsectrl.py [hunk 8]
```diff
@@ -0,0 +1,775 @@
...
+        controlnet_block = nn.Conv2d(output_channel, output_channel, kernel_size=1)
+        controlnet_block = zero_module(controlnet_block)
+        self.controlnet_down_blocks.append(controlnet_block)
+
+        for i, down_block_type in enumerate(down_block_types):
+            input_channel = output_channel
+            output_channel = block_out_channels[i]
+            is_final_block = i == len(block_out_channels) - 1
+
+            down_block = get_down_block(
```
Comment (DN6) [inline src/diffusers/models/controlnet_sparsectrl.py:298]: Let's not use `get_down_block`. Import the necessary blocks directly and use something similar to 
https://github.com/huggingface/diffusers/blob/f57b27d2ad52b7d3958a83d9ec4ce4cb6d42f615/src/diffusers/models/unets/unet_kandinsky3.py#L102

You can just use something like 

```python
if block_type == "DownBlockCrossAttnMotion":
   .....

```

File: src/diffusers/models/unets/unet_3d_blocks.py [hunk 9]
```diff
@@ -58,6 +58,7 @@ def get_down_block(
     resnet_time_scale_shift: str = "default",
     temporal_num_attention_heads: int = 8,
     temporal_max_seq_length: int = 32,
+    temporal_double_self_attention: bool = True,
```
Comment (DN6) [inline src/diffusers/models/unets/unet_3d_blocks.py:61]: We can remove this argument from here. I would import the blocks directly into the model and set the argument directly in the block.

File: src/diffusers/models/unets/unet_3d_blocks.py [hunk 10]
```diff
@@ -1574,7 +1581,7 @@ def __init__(
                     num_attention_heads=temporal_num_attention_heads,
                     in_channels=out_channels,
                     num_layers=temporal_transformer_layers_per_block[i],
-                    norm_num_groups=temporal_norm_num_groups,
```
Comment (DN6) [inline src/diffusers/models/unets/unet_3d_blocks.py:1577]: Nice catch here. Can we just remove the `temporal_norm_num_groups` argument from the `UpBlockMotion` class

Comment (a-r-r-o-w) [reply src/diffusers/models/unets/unet_3d_blocks.py:1577]: Let's do this in #8846?

File: src/diffusers/models/controlnet.py (hunk 7 above)
Comment (a-r-r-o-w) [reply src/diffusers/models/controlnet.py:833]: This was intentional, can revert if not needed. I thought there was an extra line between the comment and code by mistake

**Comment by a-r-r-o-w** [comment]
I've updated the documentation images with links from the Hub.

@DN6 The checkpoints are ready to be moved. Updated their model cards:
- SparseCtrl Scribble: https://huggingface.co/a-r-r-o-w/animatediff-sparsectrl-scribble
- SparseCtrl RGB: https://huggingface.co/a-r-r-o-w/animatediff-sparsectrl-rgb
- SparseCtrl Motion Lora: https://huggingface.co/a-r-r-o-w/animatediff-motion-lora-v1-5-3

**Comment by a-r-r-o-w** [comment]
Phew! the tests have passed - thanks for bearing with me on this one :)

**Comment by eps696** [comment]
@a-r-r-o-w thanks for the response!
alas it still does not seem to handle properly the input image, giving similar outputs no matter what.
the code is copied from "RGB single image" here https://github.com/huggingface/diffusers/pull/8897#issuecomment-2237767626 , with latest diffusers from the github main branch
animatediff_sparsectrl_rgb_single
animatediff_sparsectrl_rgb_single-2

**Comment by a-r-r-o-w** [comment]
@eps696 Have you tried these examples with the original AnimateDiff code base? The [demos](https://github.com/guoyww/AnimateDiff#demos) they show seem to have the first frame being a different darker color than remaining frames, and there are some reports as well: https://github.com/guoyww/AnimateDiff/issues/278 and https://github.com/guoyww/AnimateDiff/issues/274.

But I find it bizarre that it is not taking into account the input image at all and only following the prompt. Will debug this to find differences from our implementation vs theirs. It'd be of great help if you could check out their repo and see if you get the same incorrect behaviour like here since I recently deleted their codebase and do not have it setup at the moment + other priorities.

**Comment by eps696** [comment]
@a-r-r-o-w the difference in the keyframe color is known issue, it happened with other implementation too, maybe it's a model problem indeed. yet diffusers implementation seems to miss somehow overall motion propagation (from the keyframe over the video), which i didn't see elsewhere.

i'll check the original implementation; for now here is tests from comfyui (with the same inputs and settings):
ad_00040a
ad_00040b

**Comment by a-r-r-o-w** [comment]
@chunk-ma You mentioned [here](https://github.com/guoyww/AnimateDiff/issues/274#issuecomment-1902519273) that you found the cause of first frame being different color than rest - could you please share what changes you made?

**Comment by a-r-r-o-w** [comment]
> @a-r-r-o-w the difference in the keyframe color is known issue, it happened with other implementation too, maybe it's a model problem indeed. yet diffusers implementation seems to miss somehow overall motion propagation (from the keyframe over the video), which i didn't see elsewhere.

okay i might have overlooked some implementation details if these are the comfy results :face_in_clouds: i will take a better look soon!

**Comment by eps696** [comment]
@a-r-r-o-w just to confirm that the original AnimateDiff code also works as expected:
eps

source

**Comment by aihopper** [comment]
How high is fixing this issue in your priority list? I'd be helpful to know what to expect so we can get aligned with your roadmap.

**Comment by a-r-r-o-w** [comment]
> How high is fixing this issue in your priority list? I'd be helpful to know what to expect so we can get aligned with your roadmap.

It was high on my priority list when this was added and I gave it some time trying to look at numerical differences. I wasn't able to figure out what's wrong though, and now other important things are on my plate at the moment. Would be awesome if you'd like to take a look but I have a few more CogVideoX follow-ups before I can get back to this. I think only RGB SparseCtrl is broken, so at best it is just probably some simple internal state in the controlnet I'm not using correctly.

**Pull request merged**